### DISCUSSION TITLE
用Powerpoint寫程式? - Great Impractical Ideas in Computer Science: PowerPoint Programming

### DISCUSSION THREAD

**Discussion by lin157644** [1101-GS4719-C/Python-Community-GS4719-C#1419] [General]
# 用Powerpoint寫程式? - Great Impractical Ideas in Computer Science: PowerPoint Programming

相信各為平時都會使用Powerpoint製作簡報﹑海報吧，對我來說Powerpoint是大學生活中必不可少的工具。
本篇要與大家分享的是最近在Youtube上看到的一部影片，徹底利用Powerpoint的功能，把我們平常的簡報工具作為一種Programming的環境。

<p align="center">
  <a href="http://www.youtube.com/watch?v=_3loq22TxSc">
    <img src="http://img.youtube.com/vi/_3loq22TxSc/0.jpg">
  </a>
</p>

影片作者[Tom Wildenhain](https://www.youtube.com/@tomwildenhain9539)，是畢業自CMU的大神，他的[頻道](https://www.youtube.com/@tomwildenhain9539)中還有其他精采的應用，包括[將C語言編譯成PPT](https://youtu.be/LArkm4v5mWA)。

---

### Powerpoint programming的基本
首先，Powerpoint的超連結可以連結至其他影片，如果我們將每張投影片視為一個**狀態**，點選超連結視為**轉移條件**，我們就可以將Powerpoint類比為一台[DFA](https://zh.wikipedia.org/wiki/%E7%A1%AE%E5%AE%9A%E6%9C%89%E9%99%90%E7%8A%B6%E6%80%81%E8%87%AA%E5%8A%A8%E6%9C%BA)，在更進階的應用中，還可以利用動畫功能做到[單個投影片多個狀態](https://youtu.be/_3loq22TxSc?t=600)。

<p align="center">
  <img  height="300" src="https://github.com/NCU-GS4719-Python/Python-Community/assets/19168554/8c4a460d-96ca-4088-9485-d6408455d61b">
</p>

---

### Powerpoint圖靈機
<p align="center">
  <a href="http://www.youtube.com/watch?v=_3loq22TxSc?t=1067">
    <img src="http://img.youtube.com/vi/_3loq22TxSc/0.jpg">
  </a>
</p>

他在[這篇Paper](https://www.andrew.cmu.edu/user/twildenh/PowerPointTM/Paper.pdf)對他所設計的Powerpoint Turing machine(PPTXTM)做詳細的介紹，其中也有對Powerpoint的Turing Completeness與其限制作出解釋，並分析時間複雜度。此外，Paper也中可以找到下載簡報檔的連結，有興趣的人也可以下載來動手玩玩看。

<p align="center">
    <img src="https://github.com/NCU-GS4719-Python/Python-Community/assets/19168554/48676fbe-684b-4128-a3e9-4fadd84fd750">
</p>

<p align="center">
從Paper中對於時間複雜度的分析就可以看出來標題中的 <b>Impractical Ideas</b> 的意思
</p>

---

#### "轉化"與"手繪多邊形"

透過讓圖形在兩張投影片中含有相同文字，可以強制Powerpoint在兩個圖形之間套用轉化效果。
Powerpoint會盡力的扭曲圖形並轉換，即使是圖片也是一樣。

<p align="center">
    <img src="https://github.com/NCU-GS4719-Python/Python-Community/assets/19168554/30f3ea67-caac-4276-b41a-a2e74a0d1b15">
</p>

#### 如何應用Powerpoint內的免費3D資源
Office系列較新以x結尾的檔案本質上是一個zip檔，比如說`.pptx` `.docx` `.xlsx`。
當你在Powerpoint中插入3D圖形後，可以在zip檔 `/ppt/media` 中找到模型的檔案。
> 順道一提早期密碼保護可以透過以zip打開的方式繞過，並且密碼還是以明文儲存🤦。

#### "縮放"功能
縮放功能可以在一張簡報中嵌入另一張簡報的縮圖，透過將簡報嵌入自己，Powerpoint就會遞迴的算出碎形。

<p align="center">
    <img height="300" alt="fractal" src="https://github.com/NCU-GS4719-Python/Python-Community/assets/19168554/185e9149-7dfd-4219-9c69-29afdda0c0ed">
</p>

> 好奇為什麼要叫縮放而不叫縮圖(Thumbnail)就好

**Comment by Tzuhsuannnnn** [comment]
好特別的資訊! 雖然感覺平常不太會常用到，但超酷的
